### PR TITLE
feat: dark theme

### LINKED ISSUE #185: Day/Night mode support
It would be cool to have a button to be able to switch to darker colors. The choice should also be saved.

### LINKED ISSUE #525: Dark mode
It would be great if there was a button to change the theme to dark

### PR DESCRIPTION
**Pull request by atsuyaw**
Resolves #185, resolves #525 (relational topics: #315, #526).

Feel free to post any comments including on the default color scheme ;)

image

### CODE REVIEW

**Comment by igor-baiborodine** [comment]
@atsuyaw, thank you very much! It's a big pull request and will take me time to review.

File: v4/layouts/partials/css.html [hunk 1]
```diff
@@ -8,6 +8,10 @@
     <link rel="stylesheet" href="{{ $css.Permalink }}" integrity="{{ $css.Data.Integrity }}" media="screen">
 {{ end }}
 
+{{ $dkoptions := dict "targetPath" "dk-theme.css" "enableSourceMap" true "precision" 6 }}
+{{ $dkTheme := resources.Get "sass/dk-theme.scss" | resources.ExecuteAsTemplate "dkTheme.scss" . | toCSS $dkoptions }}
+<link disabled id="darkTheme" rel="stylesheet" href="{{ $dkTheme.Permalink }}" media="screen">
+
```
Comment (igor-baiborodine) [inline v4/layouts/partials/css.html:14]: I want to suggest updating as follows:
```html
{{ $options := dict "targetPath" "theme.css" "enableSourceMap" true "precision" 6 }}
{{ $css := resources.Get "sass/theme.scss" | resources.ExecuteAsTemplate "theme.scss" . | toCSS $options }}

{{ if isset .Site.Params "enableLightDarkMode" }}
    {{ $darkOptions := dict "targetPath" "dark-theme.css" "enableSourceMap" true "precision" 6 }}
    {{ $darkCss := resources.Get "sass/dark-theme.scss" | resources.ExecuteAsTemplate "dark-theme.scss" . | toCSS $dark-options }}
{{ end }}

{{ if or (.Site.IsServer) (eq hugo.Environment "development") }}
    <link rel="stylesheet" href="{{ $css.Permalink }}" media="screen">
    {{ if $darkCss }}
    <link disabled id="dark-theme-css" rel="stylesheet" href="{{ $darkCss.Permalink }}" media="screen">
    {{ end }}
{{ else }}
    {{ $css := $css | resources.Minify | resources.Fingerprint }}
    <link rel="stylesheet" href="{{ $css.Permalink }}" integrity="{{ $css.Data.Integrity }}" media="screen">
    {{ if $darkCss }}
    {{ $darkCss := $darkCss | resources.Minify | resources.Fingerprint }}
    <link disabled id="dark-theme-css" rel="stylesheet" href="{{ $darkCss.Permalink }}" integrity="{{ $css.Data.Integrity }}" media="screen">
    {{ end }}
{{ end }}

{{ if isset .Site.Params "css_modules" }}
    {{ range .Site.Params.css_modules }}
        <link rel="stylesheet" href="{{ . | absURL }}">
    {{ end }}
{{ end }}
```

**Review by igor-baiborodine — CHANGES_REQUESTED** [review]
There should not be duplicated `_dk-*.scss` files. The only new SCSS file should be `dark-theme.css` which is a copy of `theme.css` but with different values for customizable variables.

File: v4/layouts/partials/css.html (hunk 1 above)
Comment (atsuyaw) [reply v4/layouts/partials/css.html]: Thanks, but hugo cannot process your code.
I tried the following with a small modification, but the mode selector doesn't work correctly due to `dark-theme.css` be not loaded.
Do you have any idea?
```sass
{{ $options := dict "targetPath" "theme.css" "enableSourceMap" true "precision" 6 }}
{{ $css := resources.Get "sass/theme.scss" | resources.ExecuteAsTemplate "theme.scss" . | toCSS $options }}
{{ $darkCss := "" }}

{{ if .Site.Params.enableLightDarkMode }}
    {{ $darkOptions := dict "targetPath" "dark-theme.css" "enableSourceMap" true "precision" 6 }}
    {{ $darkCss := resources.Get "sass/dark-theme.scss" | resources.ExecuteAsTemplate "dark-theme.scss" . | toCSS $darkOptions }}
{{ end }}

{{ if or (.Site.IsServer) (eq hugo.Environment "development") }}
    <link rel="stylesheet" href="{{ $css.Permalink }}" media="screen">
    {{ if $darkCss }}
    <link disabled id="dark-theme-css" rel="stylesheet" href="{{ $darkCss.Permalink }}" media="screen">
    {{ end }}
{{ else }}
    {{ $css := $css | resources.Minify | resources.Fingerprint }}
    <link rel="stylesheet" href="{{ $css.Permalink }}" integrity="{{ $css.Data.Integrity }}" media="screen">
    {{ if $darkCss }}
    {{ $darkCss := $darkCss | resources.Minify | resources.Fingerprint }}
    <link disabled id="dark-theme-css" rel="stylesheet" href="{{ $darkCss.Permalink }}" integrity="{{ $darkCss.Data.Integrity }}" media="screen">
    {{ end }}
{{ end }}

{{ if isset .Site.Params "css_modules" }}
    {{ range .Site.Params.css_modules }}
        <link rel="stylesheet" href="{{ . | absURL }}">
    {{ end }}
{{ end }}
```

Removing`{{ if .Site.Params.enableLightDarkMode }}` and the corresponding `{{ end }}`, this code works.

Comment (atsuyaw) [reply v4/layouts/partials/css.html]: Ah, `theme.js` always needs `dark-theme.css`. So, it's difficult to accept this request.

**Comment by atsuyaw** [comment]
Thanks for helping!
Let me try to "polish" the colors prior to merging.

**Comment by atsuyaw** [comment]
@igor-baiborodine Now I made a modification on the colors. Would you check again?
Thank you!

**Comment by igor-baiborodine** [comment]
@atsuyaw, here is the Bilberry Sandbox deployed using your branch. You can compare it with the Bilberry Sandbox deployed using the latest release.
* https://deploy-preview-45--bilberry-sandbox.netlify.app/
* https://www.bilberry-sandbox.kiroule.com/
The light mode should be tested for regression. Here I spotted the first one: the line connecting the post icons. Also, the theme mode selection is not saved, and going to a different page with the dark mode selected defaults to the light mode.

image

image

**Review by igor-baiborodine — CHANGES_REQUESTED** [review]
@atsuyaw, please see my comment above.

**Comment by Lednerb** [comment]
Hello all,

thanks for your great work!

The colors are OK for me in general, but I think some adjustments should be added:

grafik

1: The lighter color coming from above will add some more contrast
2: I think the meta line should also have a bit more contrast, like the font in the article footer
3: The background is too bright, maybe more like GitHub `inline code` in the dark mode?


grafik

Additionally the search box should also be adjusted before the final merge.

And hover states on links are not readable:
grafik



----

The color theme switcher is very neat!

File: v4/layouts/partials/css.html [hunk 2]
```diff
@@ -1,15 +1,21 @@
 {{ $options := dict "targetPath" "theme.css" "enableSourceMap" true "precision" 6 }}
```
[the remark this answers is not in the document]
Comment (igor-baiborodine) [reply v4/layouts/partials/css.html:1]: No, usually when you run your site in local dev, i.e., in `development` mode, the final CSS shouldn't be minified and fingerprinted.

**Comment by igor-baiborodine** [comment]
@atsuyaw, the Bilberry Sandox was updated with your latest commits. The only issue I see now is the flickering that happens when changing pages in the dark mode.

[Screencast from 2023-11-13 05:53:09 PM.webm](https://github.com/Lednerb/bilberry-hugo-theme/assets/1027701/9f580350-2bf3-498d-9f8b-74482829190d)

File: v4/assets/sass/_articles.scss [hunk 3]
```diff
@@ -11,7 +11,7 @@ article {
         content: "";
         display: block;
         border: 11px solid transparent;
-        border-right-color: #ddd;
+        border-right-color: $con-line-colorB;
```
Comment (igor-baiborodine) [inline v4/assets/sass/_articles.scss:14]: Can the variable name be more descriptive? `con` is it for connection? `colorB` -> color-before.
The same goes for `con-line-colorA`.

**Comment by atsuyaw** [comment]
@igor-baiborodine,
Thanks for your cooperation and pointing out!
I need to modify also the giscus theme behavior with the global dark mode.

File: v4/assets/sass/dark-theme.scss [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    top: 0;
+    left: 0;
+    width: 100%;
+    height: 100%;
+    z-index: 5;
+    opacity: 0.25;
+  }
+}
+
+* {
```
Comment (igor-baiborodine) [inline v4/assets/sass/dark-theme.scss:55]: I don't think it's necessary to override the scrollbar provided by the browser. Let's keep the default one.

File: v4/README.md [hunk 5]
```diff
@@ -304,6 +304,13 @@ this will only work when the `permanentTopNav` is also set to `true`.
 
 Please note that the top navigation bar is minimized by default on mobile devices.
 
+### Light/Dark Theme Mode
+Switching between light and dark theme mode can be enabled by setting the `enableLightDarkMode` parameter to `true`:
+```toml
+[params]
+  enableLightDarkThemeMode = true
```
Comment (igor-baiborodine) [inline v4/README.md:311]: Should be `enableLightDarkTheme`.

**Review by igor-baiborodine — CHANGES_REQUESTED** [review]
The connection line should be of the same color, i.e., the same as in the part. 

image

There is still the issue of flickering when using the Chrome browser. In Firefox, it works fine.

**Comment by atsuyaw** [comment]
Here, the switching mechanism has been reconstructed.
Please try to check the improvement; the Network throttling mode in inspecting tools of Chrome was useful in my cording.

**Review by igor-baiborodine — COMMENTED** [review]
I updated the sandbox, and there is no flickering, but it's going back to the light mode when I change a page after selecting the dark mode.

https://deploy-preview-45--bilberry-sandbox.netlify.app/

File: v4/layouts/partials/check-theme-prefer.html [hunk 6]
```diff
@@ -0,0 +1,9 @@
+{{ $jsToConcat := resources.Get "js/check-theme-prefer.js" }}
+{{ $checkThemePrefer := slice $jsToConcat | resources.Concat "check-theme-prefer.js" }}
```
[the remark this answers is not in the document]
Comment (atsuyaw) [reply v4/layouts/partials/check-theme-prefer.html]: Thanks, I applied this!

**Review by igor-baiborodine — COMMENTED** [review]
I updated the Sandbox, and everything works fine, but there is a minor issue with showing `js/check-theme-prefer.js` in the top left corner.

https://deploy-preview-45--bilberry-sandbox.netlify.app/

image

**Review by atsuyaw — COMMENTED** [review]
Huuum, i saw it on the sandbox, however this issue wasn't observed in my local environment. Now i'm trying fix this line.
Let me check on the sandbox. I am sorry for inconvenience...!

**Review by igor-baiborodine — APPROVED** [review]
@atsuyaw, great job! Thank you!

**Pull request merged**